### ISSUE TITLE
Results of Cityscapes using EncNet

### ISSUE THREAD

**Issue by qiulesun** [zhanghang1989/PyTorch-Encoding#186]
I want to get results of Cityscapes dataset by training EncNet. Although you did not provide corresponding results, can I do it using directly the codes you have released, mainly involving prepare_cityscapes.py and cityscapes.py and other default hyper-parameters like lr=0.01, epoch=240?

**Comment by zhanghang1989** [comment]
You also need to set the `base-size=2048 crop-size=768`.
State-of-the-art results on Cityscapes dataset usually require a lot of tricks, such as Mapilary pretrain, OHEM loss, data balance sampling. This script can only achieve limited results.

**Comment by qiulesun** [comment]
@zhanghang1989  Will you plan to release corresponding settings to achieve them and give results under them?

**Comment by zhanghang1989** [comment]
Sorry that the top results on Cityscapes won't be provided, because I am not an expert for that dataset.

**Comment by qiulesun** [comment]
@zhanghang1989 , The Cityscapes  dataset provides 30 class labels, while most methods adopt 19 classes to evaluate their method. In order to compare with other methods, EncNet should also adopt 19 classes and the current implementation statifies that?
The question maybe too simple and I'm still looking forword to your replying if you have enough time.

**Comment by zhanghang1989** [comment]
Yes. Label is handled https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/datasets/cityscapes.py#L43

**Comment by qiulesun** [comment]
Thank you for your timely reply. Another simple but deeply troubled question will be raised.
What do the numbers in the following line mean and what are they used to do?
The line is https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/datasets/cityscapes.py#L35
        self._key = np.array([-1, -1, -1, -1, -1, -1,
                              -1, -1,  0,  1, -1, -1, 
                              2,   3,  4, -1, -1, -1,
                              5,  -1,  6,  7,  8,  9,
                              10, 11, 12, 13, 14, 15,
                              -1, -1, 16, 17, 18])


Same question also occurs at https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/datasets/pcontext.py#L35
self._mapping = np.sort(np.array([
0, 2, 259, 260, 415, 324, 9, 258, 144, 18, 19, 22,
23, 397, 25, 284, 158, 159, 416, 33, 162, 420, 454, 295, 296,
427, 44, 45, 46, 308, 59, 440, 445, 31, 232, 65, 354, 424,
68, 326, 72, 458, 34, 207, 80, 355, 85, 347, 220, 349, 360,
98, 187, 104, 105, 366, 189, 368, 113, 115]))

**Comment by zhanghang1989** [comment]
They are selected class indices.

**Comment by qiulesun** [comment]
Thank you for your patient explanation. 
I understand that. For Cityscapes, the class indices are mapping to 0:18. According to the self._classes, we can obtain self._key  (7->0, 8->1,..., background->-1). Based on the function _class_to_index, one can get mapped index (0:18) as target to guide the training rather than using their original classes (0, 7, 8, ..., 33).

**Comment by qiulesun** [comment]
@zhanghang1989 有一个不合时宜的问题想请教你。
我提交了voc2012的val set到server去评估结果，不用test set的目的是想完成一次在线评估的流程，可是几个小时过去了，server仍然没有给出预测结果（链接：http://host.robots.ox.ac.uk:8080/eval/view_submission/24739/），请问有可能是哪里出问题了？

**Comment by zhanghang1989** [comment]
You may need to ping the organizer, just shoot him an email.
The performance on validation should be exactly the same as using my script `test.py` with `--eval`.

**Comment by theevann** [comment]
Note that you also need to change the normalization transform in the `train.py` !

**Comment by zhao181** [comment]
> You also need to set the `base-size=2048 crop-size=768`.
> State-of-the-art results on Cityscapes dataset usually require a lot of tricks, such as Mapilary pretrain, OHEM loss, data balance sampling. This script can only achieve limited results.

Dear author, 
How is you best mIou result on cityscapes when using encnet? I need this for reference.

**Comment by zhanghang1989** [comment]
> > You also need to set the `base-size=2048 crop-size=768`.
> > State-of-the-art results on Cityscapes dataset usually require a lot of tricks, such as Mapilary pretrain, OHEM loss, data balance sampling. This script can only achieve limited results.
> 
> Dear author,
> How is you best mIou result on cityscapes when using encnet? I need this for reference.

EncNet doesn't have a benchmark on cityscapes, but
[CFNet](http://openaccess.thecvf.com/content_CVPR_2019/papers/Zhang_Co-Occurrent_Features_in_Semantic_Segmentation_CVPR_2019_paper.pdf) achieves 79.60% mIoU (IoU classes) on the test set.

**Comment by zhao181** [comment]
If  resnet50 backbone and 224x224 training images are used, what will be the best mIoU?


On 1/7/2020 01:44，Hang Zhang<notifications@github.com> wrote：

You also need to set the base-size=2048 crop-size=768.
State-of-the-art results on Cityscapes dataset usually require a lot of tricks, such as Mapilary pretrain, OHEM loss, data balance sampling. This script can only achieve limited results.

Dear author,
How is you best mIou result on cityscapes when using encnet? I need this for reference.

EncNet doesn't have a benchmark on cityscapes, but
CFNet achieves 79.60% mIoU (IoU classes) on the test set.

**Comment by StacyYang** [comment]
224 sounds like very limited size. Do not expect any meaningful result.

**Comment by zhao181** [comment]
But I have only one GPU with 6G memory. 
When I set image-size=512x512 and batchsize=1.0,  after one epoch(more than 50 minutes), the mIou is only 0.087. 
What's the problem? 
On 1/10/2020 23:40，Stacy Yang<notifications@github.com> wrote：

224 sounds like very limited size. Do not expect any meaningful result.

**Comment by zhao181** [comment]
I have implement a batchnorm class, but it seems it does not work. 
Can you help check the code? this is implemented for my batchsize=1 training. 


import torch
import torch.nn as nn
from torch.nn import Parameter
from  torch.nn import init
import numpy as np
from torch._jit_internal import weak_script_method


class SynBnXd(nn.Module):
def __init__(self,num_features,momentum=0.9,totaldim=2):
super(SynBnXd, self).__init__()
if(totaldim==1):
self.viewshape=(1,num_features,1)
else:
self.viewshape = (1, num_features, 1, 1)
self.num_features = num_features
self.momentum = momentum
self.weight = Parameter(torch.ones(num_features),requires_grad=True)
self.bias = Parameter(torch.zeros(num_features),requires_grad=True)
self.register_buffer('running_mean', torch.zeros(num_features))
self.register_buffer('running_var', torch.ones(num_features))
self.register_buffer('num_batches_tracked', torch.tensor(0, dtype=torch.long))
self.reset_running_stats()


def reset_running_stats(self):
self.running_var.fill_(1)
self.running_mean.zero_()
init.constant_(self.weight,1.0)
init.constant_(self.bias,0.0)
init.constant_(self.num_batches_tracked,0)




def forward(self,xdata):
if(self.training):
with torch.no_grad():
self.num_batches_tracked +=1# xdata.shape[0]
increment=1.0/float(self.num_batches_tracked.item())
momentum=np.min([1.0-increment,self.momentum])
xdata1 = xdata.clone().transpose(0, 1).contiguous()
xdata1 = xdata1.view(self.num_features, -1)
varval=torch.var(xdata1,dim=1,unbiased=False)
meanval=torch.mean(xdata1,dim=1)
self.running_mean=self.running_mean*momentum+(1.0-momentum)*meanval
self.running_var=self.running_var*momentum+(1.0-momentum)*varval
xdata.sub_(self.running_mean.view(*self.viewshape))
xdata.div_(torch.sqrt(self.running_var.view(*self.viewshape)+1.0e-5))
xdata=self.weight.view(*self.viewshape)*xdata+self.bias.view(*self.viewshape)
return xdata


def SynBn1d(num_features,momentum=0.999):
return SynBnXd(num_features,momentum=momentum,totaldim=1)




Thanks very much and best wishes. 
On 1/11/2020 03:07，zhao126<zhaoyanguo1984@126.com> wrote：
But I have only one GPU with 6G memory. 
When I set image-size=512x512 and batchsize=1.0,  after one epoch(more than 50 minutes), the mIou is only 0.087. 
What's the problem? 
On 1/10/2020 23:40，Stacy Yang<notifications@github.com> wrote：

224 sounds like very limited size. Do not expect any meaningful result.

**Comment by qiulesun** [comment]
@zhanghang1989 I used the code (https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/experiments/segmentation/test.py#L177) to generate mask files of Pascal Context. Each pixel is assigned with corresponding color presenting a label，i.e., different colors represent different labels. Can you tell me what labels different colors represent? （e.g， purple-> person）

**Comment by zhanghang1989** [comment]
The color pallete of VOC is shown here: https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/utils/pallete.py#L28

**Comment by qiulesun** [comment]
@zhanghang1989 I notice this code in your repo.  Actually, I want to know the correspondence between colors and labels.

**Comment by ygjwd12345** [comment]
I set default base-size, crop-size and the results are very terrible.
The results of encnet and deeplab model are 63.51 and 62.55.
I am going to set base-size=2048 crop-size=768 and see how it goes

**Comment by zhanghang1989** [comment]
Cityscapes data requires larger crop and base-size. The default one is for ADE20K.

**Comment by zhanghang1989** [comment]
MMSegmentation provides the results of EncNet on Cityscapes https://github.com/open-mmlab/mmsegmentation/tree/master/configs/encnet#results-and-models

This issue can be resolved now.

**Comment by ygjwd12345** [comment]
however, mmlab is too much huge and complex I can't  understand it.
